[issue: 131] jupyter-dashboards-server npm package

parente opened the pull request:
Fixes #131. See that issue for checklist of items changed.

parente commented:
@dalogsdon @jhpedemonte should be ready for review.

jhpedemonte commented on Makefile:
```diff
@@ -199,6 +190,19 @@ integration-test-auth-local: PASSWORD=testpass
+			npm install --quiet /src && \
+			test -f ./node_modules/.bin/jupyter-dashboards-server && \
+			./node_modules/.bin/jupyter-dashboards-server --help && \
+			npm uninstall --quiet jupyter-dashboards-server'
```
1. Make part of `integration-test` target?
2. Add additional integration test (or change this one) to install then run server against KG container and test loading of notebook?
3. `npm install --quiet` isn't very quiet!

parente replied:
> 1. Make part of integration-test target?

Not as written. It's just an install test. See 2.

> Add additional integration test (or change this one) to install then run server against KG container and test loading of notebook?

Yes. Was thinking about how to do this. Not sure how. Any ideas?

> 3. npm install --quiet isn't very quiet!

jhpedemonte commented on routes/api.js:
```diff
@@ -7,6 +7,14 @@ var debug = require('debug')('dashboard-proxy:server');
+} catch(e) {
+    // from install location
+    BufferList = require('../../websocket/vendor/FastBufferList');
+}
```
The `install-test` target works fine for me, but when I installed manually into a Docker container and ran, I get the following error:

```
[root@fa640848f741:/tmp# ./node_modules/.bin/jupyter-dashboards-server                          
module.js:341
    throw err;
    ^

Error: Cannot find module '../node_modules/websocket/vendor/FastBufferList'
    at Function.Module._resolveFilename (module.js:339:15)
    at Function.Module._load (module.js:290:25)
    at Module.require (module.js:367:17)
    at require (internal/module.js:16:19)
    at Object.<anonymous> (/tmp/node_modules/jupyter-dashboards-server/routes/api.js:9:18)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Module.require (module.js:367:17)
[root@fa640848f741:/tmp# ./node_modules/jupyter-dashboards-server/bin/jupyter-dashboards-server 
module.js:341
    throw err;
    ^

Error: Cannot find module '../node_modules/websocket/vendor/FastBufferList'
    at Function.Module._resolveFilename (module.js:339:15)
    at Function.Module._load (module.js:290:25)
    at Module.require (module.js:367:17)
    at require (internal/module.js:16:19)
    at Object.<anonymous> (/tmp/node_modules/jupyter-dashboards-server/routes/api.js:9:18)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Module.require (module.js:367:17)
```

parente replied:
Yes, the path to that vendor asset is finicky depending on how node lays out the deps. I'll try it manually again to see if I can reproduce and guess another way to formulate the try/catch on that vendor path.

Sidenote: is there something better we could be doing here with this vendor bit?

jhpedemonte replied:
Oh! The error is because you still have the older require code at line 9.

I'm looking at `websocket` to see if we can do something about this include. Don't like it.

parente replied:
Oh geez. That was a rebase problem. Let me fix that at least. But yes, let's find something cleaner if we can.

parente replied:
Fixed the rebase problem.

jhpedemonte commented:
Started local container like so:

```
demo-container: | build kernel-gateway-container ## Run dashboard server in a docker container
	@echo '-- Starting dashboard server container'
	$(DASHBOARD_SERVER) -it --rm \
		-v `pwd`:/src \
		--user root \
		--entrypoint /bin/bash \
		-e KERNEL_GATEWAY_URL=$(KERNEL_GATEWAY_URL) \
		-e KG_AUTH_TOKEN=$(KG_AUTH_TOKEN) \
		-e KG_BASE_URL=$(KG_BASE_URL) \
		--link $(KG_CONTAINER_NAME):$(KG_CONTAINER_NAME) \
		$(DASHBOARD_IMAGE_NAME) \
		-c 'cd /tmp && \
			npm install --quiet /src && \
			./node_modules/.bin/jupyter-dashboards-server'
```

Accessing http://192.168.99.100:3000 gives me this error:

```
Error: ENOENT: no such file or directory, open '/tmp/views/layouts/main.handlebars'
```

parente commented:
Handlebars was not configured quite right. Need to set the partials and the layouts directories if setting the overall views directory according to their doc. An abs path for all of them computed off `__dirname` works for both the source tree and npm installed location

parente commented:
Discussed with @jhpedemonte that I'll create some basic sniff tests that ensure correct server responses after npm install. That or try to re-run a subset of the integration tests on the npm installed copy.

We agreed to leave the fastbufferlist try/catch in place for now since the direct reference to it may get replaced in #97.

parente commented:
@jhpedemonte `make install-test` runs a subset of the integration tests against the npm installed copy of `jupyter-dashboards-server`. It's going to be a bit flaky because it has to wait on a timer and assume the install has completed before it expires and runs the tests (like all the other integration tests).

`make quick-install-test` is also there to simply do the npm install, run the CLI, and run the uninstall.

Should be GTG.

parente commented:
@jhpedemonte and I discussed that we'd hold this until #97 was fixed since it's likely to remove the vendor dep. Plus we're in no hurry because we're not ready to release 0.5 yet.

parente commented:
@jhpedemonte I rebased this after merging your websocket changes. I found some non-functioning integration tests along the way (upload not declaring done) and fixed those.

If you want to give this a final run through by doing `make install-test`, it should be good to merge.

jhpedemonte commented:
Looks good! Merging.
